### PR TITLE
[Owners Review] Adding support for ViUInt8 scalar and ViUInt8 array along with niFake tests

### PR DESCRIPTION
**Pull request by ni-pkhare**
### Justification

Digital Pattern has some APIs which uses `ViUInt8 scalar` and `ViUInt8[]` whose support was not included in the `mako script` earlier.
This PR adds `support` for `ViUInt8 scalar` ,`ViUInt8 as input array` and `ViUInt8 as output array`. It also includes `unit tests` in `ni_fake_service_tests.cpp`.

Link to tasks : 
- https://ni.visualstudio.com/DevCentral/_workitems/edit/1451825 
- https://ni.visualstudio.com/DevCentral/_workitems/edit/1368443

Note: The Digital Pattern APIs using these data types will be added in the upcoming PR.

### Implementation

- `ViUInt8` is basically `unsigned char` as per **IviVisaType.h** file, hence it is treated as `uint32 for scalar type` and `bytes for array type` since bytes is equivalent to string in C++.
- Since it is used as bytes in protobuf hence it is placed in the same list as that of `ViChar[]` and `ViInt8[]`. 
- Typecasting to `(ViUInt8*)` suffice our job.
for input :  `ViUInt8* an_array = (ViUInt8*)request->an_array().c_str();`
for output: 
image

- **Unit tests** are included for `ViUInt8 scalar`, `ViUInt8[] as input` and `ViUInt8[] as output`.
- Changes in **nifake metadata** is documented in **changes.md** file

### What testing has been done?

Manually ran the tests.

### CODE REVIEW

File: source/codegen/common_helpers.py [hunk 1]
```diff
@@ -13,6 +13,9 @@ def is_enum(parameter):
 def is_struct(parameter):
   return parameter["type"].startswith("struct")
 
+def is_byte_type(dataType):
+  return dataType == 'ViUInt8[]'
```
Comment (ni-pkhare) [inline source/codegen/common_helpers.py:17]: In future, if more similar DataTypes come, then it can be changed to

```
def is_byte_type(dataType) :
     return dataType in ['ViUInt8', new_data_type]
```

File: source/codegen/metadata/nifake/CHANGES.md [hunk 2]
```diff
@@ -7,6 +7,14 @@ The following functions, not originally in nimi-python metadata, were newly adde
 - `'CloseExtCal'`
 - `'BoolArrayInputFunction'`
 	- This function allows testing of ViBoolean[] input parameter
+- `'BoolArrayOutputFunction'`
+	- This function allows testing of ViBoolean[] output parameter
```
Comment (ni-pkhare) [inline source/codegen/metadata/nifake/CHANGES.md]: Note: Updating this function in `changes.md` because it was already handled sometime back.

**Review by kruthi-NI — COMMENTED** [review]
I will review after Rohan has taken a look at it.

File: source/tests/unit/ni_fake_service_tests.cpp [hunk 3]
```diff
@@ -1198,6 +1199,81 @@ TEST(NiFakeServiceTests, NiFakeService_GetAttributeViString_CallsGetAttributeViS
...
+}
+
+TEST(NiFakeServiceTests, NiFakeService_ViUInt8ArrayInputFunction_CallsViUInt8ArrayInputFunction)
+{
+  nidevice_grpc::SessionRepository session_repository;
+  std::uint32_t session_id = create_session(session_repository, kTestViSession);
+  NiFakeMockLibrary library;
+  nifake_grpc::NiFakeService service(&library, &session_repository);
+  ViInt32 number_of_elements = 3;
+  ViUInt8 expected_array[] = {0, 255, -1};
```
Comment (ni-pkhare) [inline source/tests/unit/ni_fake_service_tests.cpp]: Assigning `-1` to an `unsigned char` passes in `Windows Build` but gives compilation error in `Ubuntu Build` and `Linux-RT Build`.
So should this be considered as a valid test case or should I use any non-negative value instead of -1?
Any thoughts @doshirohan ?

File: source/tests/unit/ni_fake_service_tests.cpp [hunk 4]
```diff
@@ -1198,6 +1199,81 @@ TEST(NiFakeServiceTests, NiFakeService_GetAttributeViString_CallsGetAttributeViS
...
+}
+
+TEST(NiFakeServiceTests, NiFakeService_ViUInt8ArrayOutputFunction_CallsViUInt8ArrayOutputFunction)
+{
+  nidevice_grpc::SessionRepository session_repository;
+  std::uint32_t session_id = create_session(session_repository, kTestViSession);
+  NiFakeMockLibrary library;
+  nifake_grpc::NiFakeService service(&library, &session_repository);
+  ViInt32 number_of_elements = 3;
+  ViUInt8 an_array[] = {0, 255, -1};
```
Comment (ni-pkhare) [inline source/tests/unit/ni_fake_service_tests.cpp:1258]: Similarly here.

File: source/tests/unit/ni_fake_service_tests.cpp (hunk 3 above)
Comment (kruthi-NI) [reply source/tests/unit/ni_fake_service_tests.cpp]: @ni-pkhare , Datatype is ViUnt8 know. Why do we need to test with negative values? I think you can remove that -1. Same for other comment below

Comment (ni-pkhare) [reply source/tests/unit/ni_fake_service_tests.cpp]: @kruthi-NI , I wanted to check if it reads it as 255 or not. Thats it! Because on assigning -1 the bit representation would be 11111111 , when this bit representation is read as unsigned char then it should evaluate to 255.

Comment (kruthi-NI) [reply source/tests/unit/ni_fake_service_tests.cpp]: We are not expecting that negative values will be passed for this. Please remove this.

Comment (ni-pkhare) [reply source/tests/unit/ni_fake_service_tests.cpp]: Done.

Comment (doshirohan) [reply source/tests/unit/ni_fake_service_tests.cpp]: Agree with Kruthi. While functionally it should work the way you have mentioned, passing in negative value for UInt8 is definitely not a valid use case and need not be tested.

Comment (ni-pkhare) [reply source/tests/unit/ni_fake_service_tests.cpp]: Yes, I have removed it.

File: source/codegen/metadata/nifake/CHANGES.md (hunk 2 above)
Comment (doshirohan) [reply source/codegen/metadata/nifake/CHANGES.md]: This function was not introduced by us. It was part of nimi-python metadata, Remove this line.

File: source/codegen/templates/service.cpp.mako [hunk 5]
```diff
@@ -37,6 +37,20 @@ namespace ${config["namespace_component"]}_grpc {
   {
   }
 
+% if common_helpers.has_bytes_type_param(functions):
+  template <typename T1, typename T2>
+  void ${service_class_prefix}Service::Copy(const std::vector<T1>& input, T2 output)
+  {
+    output->insert(0, (size_t)input.size(), '\0');
```
Comment (doshirohan) [inline source/codegen/templates/service.cpp.mako]: Why do you need to overwrite everything with `\0`? It will be overwritten with actual data anyway.

File: source/codegen/templates/service.cpp.mako [hunk 6]
```diff
@@ -37,6 +37,20 @@ namespace ${config["namespace_component"]}_grpc {
   {
   }
 
+% if common_helpers.has_bytes_type_param(functions):
+  template <typename T1, typename T2>
+  void ${service_class_prefix}Service::Copy(const std::vector<T1>& input, T2 output)
```
Comment (doshirohan) [inline source/codegen/templates/service.cpp.mako]: You can't assume type of `T2`, so it is wrong to call `insert` and `output->data()` on it. Also, you definitely shouldn't typecast it to `T1*` on line 45 below. Once we add support for `VIUint16[]`, for input type parameter -> T1 would be 32 bit array and T2 would be 16 bit array. Typecasting would lead to additional elements in array. You should use `T2 *` for data type of `output` and do `output[i] = item` as we discussed earlier last week.

If you had to change the approach for using same copy function with input and output parameters, look at how ViBoolean handles copying of input data [here in service_helpers.mako](https://github.com/doshirohan/grpc-device/blob/main/source/codegen/templates/service_helpers.mako#L172) and see how that can be tweaked to work with non-32-bit integers. You would probably just need to change the lambda used for copying the elements to ` [](auto x) { return (${c_type_underlying_type})x; }`. Merge with latest feature branch before you proceed since some of these checks could potentially be merged with ViBoolean's handling.

File: source/codegen/templates/service_helpers.mako [hunk 7]
```diff
@@ -192,7 +192,7 @@ one_of_case_prefix = f'{namespace_prefix}{function_name}Request::{PascalFieldNam
 % elif c_type == 'ViSession':
       auto ${parameter_name}_grpc_session = ${request_snippet};
       ${c_type} ${parameter_name} = session_repository_->access_session(${parameter_name}_grpc_session.id(), ${parameter_name}_grpc_session.name());\
-% elif c_type == 'ViInt32[]' or c_type == 'ViAddr[]':
+% elif c_type == 'ViInt32[]' or c_type == 'ViAddr[]' or common_helpers.is_bytes_type(c_type):
       auto ${parameter_name} = const_cast<${c_type_pointer}>(reinterpret_cast<const ${c_type_pointer}>(${request_snippet}.data()));\
```
Comment (doshirohan) [inline source/codegen/templates/service_helpers.mako]: Do not `reinterpret` this. It won't work when we support ViUInt16 which is represented by 32 bit data on wire. Try using similar approach to ViBoolean[] as I mentioned in my other comment in `service.cpp.mako` file.

File: source/codegen/common_helpers.py [hunk 8]
```diff
@@ -97,6 +100,14 @@ def has_viboolean_array_param(functions):
         return True
   return False
 
+def has_bytes_type_param(functions):
```
Comment (doshirohan) [inline source/codegen/common_helpers.py]: Why did you make it specific to `bytes` type? We should make it generic enough to handle any non-32 bit integers even though we are not updating it to support ViUInt16[] yet. Change name of this and helper on line 16 above to `has_non_32_bit_integer`

**Review by doshirohan — CHANGES_REQUESTED** [review]
Update the `Copy` function as mentioned in my comments to support any non 32 bit integers.

File: source/codegen/metadata/nifake/CHANGES.md (hunk 2 above)
Comment (ni-pkhare) [reply source/codegen/metadata/nifake/CHANGES.md]: Oh I made a mistake. I have removed it.

File: source/codegen/templates/service.cpp.mako (hunk 5 above)
Comment (ni-pkhare) [reply source/codegen/templates/service.cpp.mako]: The latest commit has used `resize` function instead of `insert` function. Could you pl have a look at it?

File: source/codegen/templates/service.cpp.mako (hunk 6 above)
Comment (ni-pkhare) [reply source/codegen/templates/service.cpp.mako]: I have let it to be `T2 output` only as while calling the Copy function I am passing typecasted pointer, `Copy(an_array, (ViUInt8*)response->mutable_an_array()->data());` 
So now in place of T2 (ViUInt8*) will be taken and I can iterate over the output. The tests are working fine with this.

File: source/codegen/templates/service_helpers.mako (hunk 7 above)
Comment (ni-pkhare) [reply source/codegen/templates/service_helpers.mako]: I can handle ViUInt16[] by small changes in the current implementation itself. I can do it in subsequent PRs whenever any API comes which uses this parameter. The latest commit is not using this line, I have changed it to use c_str thing.

File: source/codegen/common_helpers.py (hunk 8 above)
Comment (ni-pkhare) [reply source/codegen/common_helpers.py]: I can handle ViUInt16[] by small changes in the current implementation itself. I can do it in subsequent PRs whenever any API comes which uses this parameter.

File: source/codegen/templates/service.cpp.mako (hunk 6 above)
Comment (doshirohan) [reply source/codegen/templates/service.cpp.mako]: If you use `T2` as type in the template, you can't really do `output[i]` since from template perspective `T2` need not be a pointer. Users can pass `Int32` as T2 and it would be valid syntactically. By marking it as `T2*`, template explicitly says it expects pointer to data as second parameter.

Comment (ni-pkhare) [reply source/codegen/templates/service.cpp.mako]: Yes, now it seems logical to me why T2* is to be used because we want this Copy function to only expect pointer type as 2nd parameter so that we can use it to iterate over output. Thanks, I have made the change.

File: source/codegen/common_helpers.py [hunk 9]
```diff
@@ -13,6 +13,9 @@ def is_enum(parameter):
 def is_struct(parameter):
   return parameter["type"].startswith("struct")
 
+def is_bytes_type(dataType):
+  return dataType == 'ViUInt8[]'
```
Comment (doshirohan) [inline source/codegen/common_helpers.py]: Add similar comment as `has_bytes_type_param` here as well. It is not obvious from implementation why `ViUint8[]` is bytes type. Adding explicit comment would make it easier to understand.

File: source/codegen/metadata/nifake/functions.py [hunk 10]
```diff
@@ -2545,5 +2545,71 @@
             }
         ],
         'returns': 'ViStatus'
+    },
```
Comment (doshirohan) [inline source/codegen/metadata/nifake/functions.py:2612]: Looks like functions in this file are added in sorted order. Insert these new functions at appropriate locations.

File: source/codegen/service_helpers.py [hunk 11]
```diff
@@ -18,6 +18,8 @@ def create_args(parameters):
         result = f'{result}({type_without_brackets}*){parameter_name}.data(), '
       elif parameter['type'] == 'ViBoolean[]':
         result = f'{result}{parameter_name}.data(), '
+      elif is_output and is_array and common_helpers.is_bytes_type(parameter['type']):
```
Comment (doshirohan) [inline source/codegen/service_helpers.py]: As I had mentioned earlier, we can merge codegen for this with the one on line 16 above. You could either `or` `is_bytes_type` with `is_string_arg`. Looking through the code, codegen for both seems pretty much same except `resize` on response object is not being done for `is_string_arg` during `initialize_output_params`.

You should definitely move `ViInt8[]` check from `is_string_arg` to `is_bytes_type` since that way we would also support input/output array types for `ViInt8[]`.

After merging these 2 checks, `is_array` condition probably can be skipped since "bytes" type check implies array type.

File: source/codegen/templates/service_helpers.mako [hunk 12]
```diff
@@ -164,7 +164,7 @@ ${initialize_standard_input_param(function_name, parameter)}\
       ${c_type} ${parameter_name} = ${request_snippet}.c_str();\
 % elif c_type == 'ViString' or c_type == 'ViRsrc':
       ${c_type} ${parameter_name} = (${c_type})${request_snippet}.c_str();\
-% elif c_type == 'ViInt8[]' or c_type == 'ViChar[]':
+% elif c_type == 'ViInt8[]' or c_type == 'ViChar[]' or common_helpers.is_bytes_type(c_type):
```
Comment (doshirohan) [inline source/codegen/templates/service_helpers.mako]: You can replace this whole statement with `is_string_arg(c_type) or is_bytes_type(c_type)`. Since `is_string_arg` accepts `parameter` right now, just go ahead and update it to instead accept type and rename to `is_string_type` to be consistent with `is_bytes_type` naming.

File: source/codegen/templates/service_helpers.mako [hunk 13]
```diff
@@ -164,7 +164,7 @@ ${initialize_standard_input_param(function_name, parameter)}\
...
       ${c_type_pointer} ${parameter_name} = (${c_type[:-2]}*)${request_snippet}.c_str();\
```
Comment (doshirohan) [inline source/codegen/templates/service_helpers.mako]: (${c_type[:-2]}*) => (${c_type_pointer})

File: source/tests/unit/ni_fake_service_tests.cpp [hunk 14]
```diff
@@ -1198,6 +1199,81 @@ TEST(NiFakeServiceTests, NiFakeService_GetAttributeViString_CallsGetAttributeViS
...
+}
+
+TEST(NiFakeServiceTests, NiFakeService_ViUInt8ArrayOutputFunction_CallsViUInt8ArrayOutputFunction)
+{
+  nidevice_grpc::SessionRepository session_repository;
+  std::uint32_t session_id = create_session(session_repository, kTestViSession);
+  NiFakeMockLibrary library;
+  nifake_grpc::NiFakeService service(&library, &session_repository);
+  ViInt32 number_of_elements = 3;
+  ViUInt8 an_array[] = {0, 127, 255};
```
Comment (doshirohan) [inline source/tests/unit/ni_fake_service_tests.cpp]: Good call using values from throughout UInt8 range. Jonathan had suggested in another PR to specify upper value in hex representation so it is clear that we are using max 8 bit value, like `0xFF`. Make that change here as well.

File: source/tests/unit/ni_fake_service_tests.cpp [hunk 15]
```diff
@@ -1198,6 +1199,81 @@ TEST(NiFakeServiceTests, NiFakeService_GetAttributeViString_CallsGetAttributeViS
...
+}
+
+TEST(NiFakeServiceTests, NiFakeService_ViUInt8ArrayInputFunction_CallsViUInt8ArrayInputFunction)
+{
+  nidevice_grpc::SessionRepository session_repository;
+  std::uint32_t session_id = create_session(session_repository, kTestViSession);
+  NiFakeMockLibrary library;
+  nifake_grpc::NiFakeService service(&library, &session_repository);
+  ViInt32 number_of_elements = 3;
+  ViUInt8 expected_array[] = {0, 127, 255};
```
Comment (doshirohan) [inline source/tests/unit/ni_fake_service_tests.cpp]: Same as comment below, represent `255` as `0xFF`.

File: source/codegen/templates/service_helpers.mako [hunk 16]
```diff
@@ -263,6 +267,8 @@ one_of_case_prefix = f'{namespace_prefix}{function_name}Request::{PascalFieldNam
         response->set_${parameter_name}(${parameter_name});
 %     elif common_helpers.is_struct(parameter) or parameter['type'] == 'ViBoolean[]':
         Copy(${parameter_name}, response->mutable_${parameter_name}());
+%     elif common_helpers.is_bytes_type(parameter['type']):
+        Copy(${parameter_name}, (${underlying_param_type}*)response->mutable_${parameter_name}()->data());
```
Comment (doshirohan) [inline source/codegen/templates/service_helpers.mako]: We probably don't need `->data()` for second parameter. I remember this comment from ViBoolean[] support. Can you remove that and check if it still works (tests pass)?

File: source/codegen/service_helpers.py (hunk 11 above)
Comment (doshirohan) [reply source/codegen/service_helpers.py]: You can look at `Export\ImportAttributeConfigurationBuffer` APIs for `ViInt8[]` and `GetAStringOfFixedMaximumSize` for `ViChar[]` in nifake_service.cpp to check how codegen for them looks like and compare that with ViUint8[] to check for differences.

**Review by doshirohan — CHANGES_REQUESTED** [review]
There is still minor cleanup with merging of `is_string_arg` and `is_bytes_type` related codegen. I want to look at that cleanup before pushing for owners. Rest of the changes seem fine.

File: source/codegen/templates/service.cpp.mako (hunk 5 above)
Comment (ni-pkhare) [reply source/codegen/templates/service.cpp.mako]: Resolving based on our recent discussion.

File: source/codegen/templates/service.cpp.mako (hunk 6 above)
Comment (ni-pkhare) [reply source/codegen/templates/service.cpp.mako]: Resolving based on our recent discussion.

File: source/codegen/common_helpers.py (hunk 8 above)
Comment (ni-pkhare) [reply source/codegen/common_helpers.py]: Resolving based on our recent discussion.

**Comment by ni-pkhare** [comment]
> There is still minor cleanup with merging of `is_string_arg` and `is_bytes_type` related codegen. I want to look at that cleanup before pushing for owners. Rest of the changes seem fine.

Based on our recent discussion, the whole conversation is resolved.

File: source/codegen/templates/service_helpers.mako [hunk 17]
```diff
@@ -243,6 +243,7 @@ one_of_case_prefix = f'{namespace_prefix}{function_name}Request::{PascalFieldNam
 % for parameter in output_parameters:
 <%
   parameter_name = common_helpers.camel_to_snake(parameter['cppName'])
+  underlying_param_type = common_helpers.get_underlying_type_name(parameter["type"])
```
Comment (doshirohan) [inline source/codegen/templates/service_helpers.mako]: This variable is no longer required. You can revert change on this line.

**Review by doshirohan — APPROVED** [review]

**Comment by doshirohan** [comment]
suggestion/opinion: You have used backticks to highlight keywords in PR description which makes it bit difficult to read. Typically they are used for marking `code` and not keywords. You can use it to highlight function names, data types, or code snippet.

File: source/codegen/templates/service_helpers.mako (hunk 17 above)
Comment (ni-pkhare) [reply source/codegen/templates/service_helpers.mako]: Done. Forgot to remove it earlier.

**Review by reckenro — APPROVED** [review]

**Review by jonlugoatx — APPROVED** [review]
